src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java

```diff
@@ -685,36 +706,137 @@ public void createDirectory(Path dir,
+            throw new IOException(e);
         }
-        if (Files.isDirectory(s3Path))
+
+        // we delete the two objects (sometimes exists the key '/' and sometimes not)
+        keys = batch.stream()
+                    .map(s3Path -> ObjectIdentifier.builder().key(s3Path.getKey() + '/').build())
+                    .collect(Collectors.toList());
+        multiObjectDeleteRequest = DeleteObjectsRequest.builder()
+                                                       .bucket(bucketName)
+                                                       .delete(Delete.builder()
+                                                                     .objects(keys)
+                                                                     .build())
+                                                       .build();
+
+        try
         {
-            try (final DirectoryStream<Path> stream = Files.newDirectoryStream(s3Path))
+            client.deleteObjects(multiObjectDeleteRequest);
+        }
+        catch (SdkException e)
+        {
+            throw new IOException(e);
+        }
+    }
+
+    private LinkedList<Deque<S3Path>> getPathsByBatch(S3Path path)
+            throws IOException
+    {
+        LinkedList<S3Path> allPaths = new LinkedList<>();
+        visitAllFiles(path, allPaths);
+
+        return splitByBatchWithOrder(allPaths, MAX_OBJECT_PER_REQUEST);
+    }
+
+    private <T> LinkedList<Deque<T>> splitByBatchWithOrder(List<T> list,
+                                                           int maxSize)
+    {
+        LinkedList<Deque<T>> pathsByBatch = new LinkedList<>();
+
+        if (!list.isEmpty())
+        {
+            Deque<T> deque = new ArrayDeque<>();
+            for (T t : list)
             {
-                if (stream.iterator().hasNext())
+                if (deque.size() < maxSize)
+                {
+                    deque.push(t);
+                }
+                else
                 {
-                    throw new DirectoryNotEmptyException("the path: " + this + " is a directory and is not empty");
+                    pathsByBatch.add(deque);
+                    deque = new ArrayDeque<>();
                 }
             }
+            pathsByBatch.add(deque); // add the last batch
+
         }
 
-        String key = s3Path.getKey();
-        String bucketName = s3Path.getFileStore().name();
-        final S3Client client = s3Path.getFileSystem().getClient();
+        return pathsByBatch;
+    }
 
-        DeleteObjectRequest request = DeleteObjectRequest.builder().bucket(bucketName).key(key).build();
-        client.deleteObject(request);
+    private void visitAllFiles(Path path,
+                               LinkedList<S3Path> paths)
+            throws IOException
+    {
+        S3Path s3Path = toS3Path(path);
+        if (Files.notExists(s3Path))
+        {
+            LOGGER.warn("Deleting " + s3Path + " was skipped because the path was not found.");
```

Comment:
Format specifiers should be used instead of string concatenation.

```suggestion
            LOGGER.warn("Deleting {} was skipped because the path was not found.", s3Path);
```